pyam/utils.py

```diff
@@ -94,13 +94,30 @@ def write_sheet(writer, name, df, index=False):
         writer.sheets[name].set_column(i, i, width)  # assumes xlsxwriter as engine
 
 
+def get_excel_file_with_kwargs(path, **kwargs):
+    """Return a `pandas.ExcelFile` and a dict of unused kwargs.
+
+    When reading an Excel file, this function finds keyword arguments that
+    should be passed to `pandas.ExcelFile`, and returns a `pandas.ExcelFile`
+    instance along with the remaining keyword arguments (which presumably
+    will be used for other purposes by the calling function).
+    """
+    EXCEL_FILE_KWS = ('engine', 'storage_options', 'engine_kwargs')
```

Comment:
```suggestion
    EXCEL_FILE_KWS = ('engine', 'storage_options', 'engine_kwargs', 'header', 'usecols', 'skiprows', 'nrows')
```

Reply:
And you could move that variable to the top of the file...

Reply:
Thanks, but that shouldn't be needed. I think you are confusing kwargs to `pandas.read_excel` and `pandas.ExcelFile` here. It's only the latter that needed to be addressed by this PR (and included in `EXCEL_FILE_KWS`). But since there was no pre-existing test for kwargs to `pandas.read_excel`, or correct handling of both simultaneously, I included that too.

`header`, `usecols`, `skiprows` and `nrows` are all arguments to `pandas.read_excel` and handled in that function, not passed on to `pandas.ExcelFile` (which does not accept those keywords). The code before this PR already supported keywords that are handled by `pandas.read_excel` and not passed to `pandas.ExcelFile`. The problem was the specific arguments that need to be passed to `pandas.ExcelFile` (engine, storage_options, engine_kwargs). `pyam.utils.read_pandas` passed its kwargs on to `pandas.read_excel`, but it first created a `pandas.ExcelFile` instance without any kwargs. pandas would then raise an exception, because it got both a pre-created `pandas.ExcelFile` instance and duplicate kwargs that are only accepted when `pandas.read_excel` is called with a Path and creates its own `pandas.ExcelFile` instance from that.

The `sheet_name` argument here is not ignored. If it were, the test would fail (I checked that first by leaving it out and confirming that the test indeed failed). I included it precisely to check that `pandas.read_excel` kwargs that are *not* passed to `pandas.ExcelFile` are not ignored. The xlsx file that is loaded in the test does not contain any sheet named `data` or otherwise recognized by pyam, so it relies on `sheet_name` being specified.

Reply:
You are of course right, I was trying to review the PR in between two minutes and didn't take the necessary time, sorry. I overlooked that you had added a new test_df.xlsx file.